docs/user_guide/ml.md

```diff
@@ -1,8 +1,254 @@
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def train
+:end-at: return
+```
+### Good practice for training:
+It is considered to be a good practice to define helper functions for model building and column preprocessing and call them in {py:meth}`~columnflow.ml.MLModel.train`.
+The reasoning behind this design decision is that in {py:meth}`~columnflow.ml.MLModel.train`, you want to focus more on the definition of the actual trainings loop.
+Another reason is that especially the preparation of events, can take a large amount of code lines, making it messy to debug.
+In the following example it is shown how to define these helper functions.
+First of all one needs a function to handle the opening and combination of all parquet files.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def open_input_files
+:end-at: return all_events
+```
+
+In `prepare_events` we use the combined awkward and filter out all columns we are not interested in during the training, to stay lightweight.
+Next we split the remaining columns into input and target column and bring these columns into the correct shape, data type, and also use certain preprocessing transformation.
+At the end we transform the awkward array into a Tensorflow tensor, something that our machine learning model can handle.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def prepare_events
+:end-at: return tf.convert_to_tensor
+```
+The actual building of the model is also handled by a separate function.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def build_model
+:end-at: return model
+```
+With this little example one can see why it is of good practice to separate this process into small chunks.
+
+## evaluate:
+Within {py:meth}`~columnflow.ml.MLModel.train` one defined the trainings process of the ML model, while in {py:meth}`~columnflow.ml.MLModel.evaluate` its evaluation is defined.
+The corresponding `task` is {py:class}`~columnflow.tasks.ml.MLEvaluation` and depends on {py:class}`~columnflow.tasks.ml.MLTraining` and will therefore trigger a training if no training was performed before.
+
+For each fold of the k-folds a neural network model is trained and can be accessed from `models`.
+The actual loading, of the trained model stored in the list, is defined in {py:meth}`~columnflow.ml.MLModel.open_model` function.
+
+The awkward array `events` is loaded in chunks and contains the merged events of all folds.
+To filter out the test set, create a mask with `fold_indices`.
+
+If you want to preserve columns and write them out into a parquet file, append the columns to `events` using `set_ak_column` and return `events`.
+All columns not present in {py:meth}`~columnflow.ml.MLModel.produces` are then filtered out.
+
+In the following example, the models prediction as well as the number of muons and electrons are saved, all the other columns in events are thrown away, since they are not present in `use`:
```

Comment:
> ...thrown away, since they are not present in `use`.

Don't you mean `produce`?

Reply:
yes :(
```suggestion
In the following example, the models prediction as well as the number of muons and electrons are saved, all the other columns in events are thrown away, since they are not present in `produce`:
```